accepted/0000-workspaces.md

```diff
@@ -0,0 +1,479 @@
+    "workspaces": {
+        "packages": [
+            "packages/*"
+        ]
+    }
+}
+```
+
+`package.json` shorthand example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": [
+        "packages/*"
+    ]
+}
+```
+
+
+### 2. Installing dependencies across child packages
+
+Change `npm install` (arborist) behavior to make it install every nested package by default once a valid configuration is defined.
+
+Arborist should also be aware of all nested workspaces and correctly link to an internal workspace should it match the required semver version of an expected dependency anywhere in the installing tree. e.g:
+
+```
+// Given this package.json structure:
+├── package.json { "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "dependencies": { "dep-b": "^1.0.0" } }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm install
+
+// Results in this symlinking structure:
+├── node_modules
+│   ├── dep-a -> ./dep-a
+│   └── dep-b -> ./dep-b
+├── dep-a
+│   └── node_modules
+│       └── dep-b -> ../../../node_modules/dep-b
+└── dep-b
+```
+
+NOTE: The final implementation of the symlinking structure might differ from the examples laid out here (which are trying to be as illustrative as possible). Arborist might end up resolving all symlinks up to the actual place instead of creating a chain of symlinks.
+
+### 3. Run commands across all child packages
+
+Make npm commands workspace-aware, so that running a command tries to run it within nested packages as long as a workspaces field is defined in `package.json`.
+
+Only a subset of npm commands are to be supported:
+
+- `fund` List funding info for all packages
+- `ls` List all packages
+- `outdated` List outdated deps for all packages
+- `publish` Publishes all packages
+- `run-script` Run arbitrary scripts for all packages
+- `rebuild` Rebuild all packages
+- `restart`
+- `start`
+- `stop`
+- `test` Run tests in all packages
+- `update` Update a dependency in all packages
+- `version` Bump versions for all packages
+- `view` View registry info for all packages
+
+A new config value should be introduced in order to allow for filtering out a subset of the packages in which to run these commands. e.g: `--filter`
+
+#### Test example:
+
+```
+├── package.json { "name": "foo", "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "version": "1.0.0" }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm test
+
+> foo@1.0.0 test /Users/username/foo
+> echo "Error: no test specified" && exit 1
+
+Error: no test specified
+npm ERR! Test failed.  See above for more details.
+
+> dep-a@1.0.0 test /Users/username/foo/dep-a
+> done
+
+> dep-b@1.3.1 test /Users/username/foo/dep-b
+> done
+```
+
+
+### 4. Publishing workspaces
+
+A workspace may not be published to the registry and for all publishing purposes having a valid `"workspace"` entry in a `package.json` is going to be the equivalent of `"private": true`.
```

Comment:
This seems like a needless restriction and blocks some interesting use cases. I can imagine a developer experience which involves teams publishing workspaces as a mechanism for sharing a setup.  You could just specify that installing a published package with a workspaces key will do nothing, and then let userland tooling extend this if it wants.

Reply:
@isaacs 

> Do we just strip off the workspaces declaration when installing

I would think for a use case like sharing a setup you would not strip it, only ignore it.

> Do the workspace packages get listed as dependencies?

No, as with other workspace implementations you only get it when developing locally with it as the top level package.  So to publish a package it must define all that it requires when installing from the registry.

Reply:
So if it's not a top of tree module (root or link target, not a deep in node_modules) then the workspaces object is irrelevant? That could work I guess.

Reply:
@wesleytodd this is looking interesting 👍 assuming there's no technical limitations that  might prevent it - I'm onboard with the suggestions

> You could just specify that installing a published package with a workspaces key will do nothing, and then let userland tooling extend this if it wants.

just to clarify, in this case there's no filtering out mechanism at the moment of publishing a top-level workspace 🤔 meaning all the contents of nested packages within it get published along with that top-level package?

Reply:
I think it's the right call to make a `"workspaces"` field in package.json equivalent to `"private": true`. "Sharing configuration" (for an identical workspace, what?) should be a separate concern.

Reply:
> meaning all the contents of nested packages within it get published along with that top-level package?

My Yeah you would use the normal methods for specifying which files make it into the tarball if you wanted different behavior.

> Sharing configuration" (for an identical workspace, what?) should be a separate concern.

Just wondering, why do you think this?  To me this clause of "workspaces means private" makes it a concern here.  I am fine having that as a separate conversation as long as this clause does not make it into the implementation.

Reply:
@wesleytodd The manifest that coordinates the workspaces, as I've commented elsewhere, should not be conflated with a manifest that specifies a publishable package. The fact that Lerna _can_ do that is an accident of history, and does not represent what I believe to be a robust pattern for managing multi-package repositories.

Reply:
Publishing it is a distribution mechanism, and the registry is the distribution mechanism in our ecosystem. If folks find novel ways to use this as both a distribution mechanism and a tool to build better developer experience, GREAT!  I don't see this needless restriction gaining us anything at this time.

While it might not not help your monorepo-centric tool doesn't mean that it cannot help other types of setups.  Do you think that your worries hold up when you move outside of a monorepo?  Do you have suggestions for supporting multi project workspaces outside of a monorepo?

My single biggest concern for this implementation, that it not be tightly coupled to the monorepo setup.  If npm just follows yarn and lerna on this I will continue *not* to be able to use them for my main use cases.  Most OSS projects are a collection of projects in their own repos.  Some popular ones outgrew the tooling (a separate problem I think we can solve), but the vast majority have not.  If we do not make this a top priority for workspaces in npm I think we will have missed an opportunity to appeal to a much larger audience.